be/src/exec/hash_join_components.cpp

```diff
@@ -155,4 +400,387 @@ ChunkPtr SingleHashJoinBuilder::convert_to_spill_schema(const ChunkPtr& chunk) c
     return _ht.convert_to_spill_schema(chunk);
 }
 
+enum class CacheLevel { L2, L3, MEMORY };
+
+class AdaptivePartitionHashJoinBuilder final : public HashJoinBuilder {
+public:
+    AdaptivePartitionHashJoinBuilder(HashJoiner& hash_joiner);
+    ~AdaptivePartitionHashJoinBuilder() override = default;
+
+    void create(const HashTableParam& param) override;
+
+    void close() override;
+
+    void reset(const HashTableParam& param) override;
+
+    Status do_append_chunk(const ChunkPtr& chunk) override;
+
+    Status build(RuntimeState* state) override;
+
+    bool anti_join_key_column_has_null() const override;
+
+    int64_t ht_mem_usage() const override;
+
+    void get_build_info(size_t* bucket_size, float* avg_keys_per_bucket) override;
+
+    size_t get_output_probe_column_count() const override;
+    size_t get_output_build_column_count() const override;
+
+    void visitHt(const std::function<void(JoinHashTable*)>& visitor) override;
+
+    std::unique_ptr<HashJoinProberImpl> create_prober() override;
+
+    void clone_readable(HashJoinBuilder* builder) override;
+
+    ChunkPtr convert_to_spill_schema(const ChunkPtr& chunk) const override;
+
+private:
+    size_t _estimated_row_size(const HashTableParam& param) const;
+    size_t _estimated_probe_cost(const HashTableParam& param) const;
+    template <CacheLevel T>
+    size_t _estimated_build_cost(size_t build_row_size) const;
+    void _adjust_partition_rows(size_t build_row_size);
+
+    void _init_partition_nums(const HashTableParam& param);
+    Status _convert_to_single_partition();
+    Status _append_chunk_to_partitions(const ChunkPtr& chunk);
+
+private:
+    std::vector<std::unique_ptr<SingleHashJoinBuilder>> _builders;
+
+    size_t _partition_num = 0;
+    size_t _partition_join_min_rows = 0;
+    size_t _partition_join_max_rows = 0;
+
+    size_t _probe_estimated_costs = 0;
+
+    size_t _fit_L2_cache_max_rows = 0;
+    size_t _fit_L3_cache_max_rows = 0;
+
+    size_t _L2_cache_size = 0;
+    size_t _L3_cache_size = 0;
+
+    size_t _pushed_chunks = 0;
+};
+
+AdaptivePartitionHashJoinBuilder::AdaptivePartitionHashJoinBuilder(HashJoiner& hash_joiner)
+        : HashJoinBuilder(hash_joiner) {
+    const size_t DEFAULT_L2_CACHE_SIZE = 1 * 1024 * 1024;
```

Comment:
static const

Reply:
I replaced them to `static constexpr size_t DEFAULT_L2_CACHE_SIZE = 1 * 1024 * 1024;`